examples/advanced/quantum_volume.py

```diff
@@ -0,0 +1,156 @@
+
+    The generated circuit consists of `depth` layers of random qubit
+    permutations followed by random two-qubit gates that are sampled from the
+    Haar measure on SU(r).
+
+    Args:
+        num_qubits: The number of qubits in the generated circuit.
+        depth: The number of layers in the circuit.
+        random_state: A way to seed the RandomState.
+
+    Returns:
+        The generated circuit.
+    """
+    # Setup the circuit and its qubits.
+    qubits = cirq.LineQubit.range(num_qubits)
+    circuit = cirq.Circuit()
+    if random_state is None:
+        random_state = np.random.RandomState()
+
+    # For each layer.
+    for _ in range(depth):
+        # Generate uniformly random permutation Pj of [0...n-1]
+        perm = random_state.permutation(num_qubits)
+
+        # For each consecutive pair in Pj, generate Haar random SU(4)
+        # Decompose each SU(4) into CNOT + SU(2) and add to Ci
+        for k in range(0, num_qubits - 1, 2):
+            permuted_indices = [int(perm[k]), int(perm[k + 1])]
+            special_unitary = cirq.testing.random_special_unitary(
+                4, random_state=random_state)
+
+            # Convert the decomposed unitary to Cirq operations and add them to
+            # the circuit.
+            circuit.append(
+                cirq.two_qubit_matrix_to_operations(qubits[permuted_indices[0]],
+                                                    qubits[permuted_indices[1]],
+                                                    special_unitary,
+                                                    allow_partial_czs=False))
+
+    # Don't measure all of the qubits at the end of the circuit because we will
+    # need to classically simulate it to compute its heavy set.
+    return circuit
+
+
+def compute_heavy_set(circuit: cirq.Circuit) -> List[str]:
+    """Classically compute the heavy set of the given circuit.
+
+    The heavy set is defined as the output bit-strings that have a greater than
+    median probability of being generated.
+
+    Args:
+        circuit: The circuit to classically simulate.
+
+    Returns:
+        A list containing all of the heavy bit-string results.
+    """
+    # Classically compute the probabilities of each output bit-string through
+    # simulation.
+    simulator = cirq.Simulator()
+    results = cast(cirq.WaveFunctionTrialResult,
+                   simulator.simulate(program=circuit))
+    # Compute the median probability of the output bit-strings.
+    median = statistics.median(results.state_vector())
+
+    # The output wave function is a vector from the result value (big-endian) to
+    # the probability of that bit-string. Compute a format string converts the
+    # given index to the corresponding qubit string.
+    format_str = '{0:0%sb}' % len(circuit.all_qubits())
+    # Return all of the bit-strings that have a probability greater than the
+    # median.
+    return [
+        format_str.format(idx)
```

Comment:
We (or at least I) generally don't like turning bit-strings into character-strings. Typically we leave them as 2d numpy integer arrays

```python
hog_bitstrings = np.array([
  [1, 0, 1, 1],
  [0, 0, 0, 0],
  ...
])
```

In this case, it might be more useful to leave them as integers (ie `idx`) to make the set-testing easier when you draw samples and want to know if they're in the HOG set. Up to you

Reply:
Sure, makes sense. Done.